runtime/src/main/scala/akka/grpc/internal/Marshallers.scala

```diff
@@ -18,19 +18,33 @@ abstract class BaseMarshaller[T](val protobufSerializer: ProtobufSerializer[T])
     extends io.grpc.MethodDescriptor.Marshaller[T]
     with WithProtobufSerializer[T] {
   override def parse(stream: InputStream): T = {
-    val baos = new ByteArrayOutputStream(math.max(64, stream.available()))
-    val buffer = new Array[Byte](32 * 1024)
+    val buffer =
+      new Array[Byte](stream match {
+        case k: KnownLength => math.max(0, k.available()) // No need to oversize this if we already know the size
+        case _              => 32 * 1024
+      })
 
     // Blocking calls underneath...
     // we can't avoid it for the moment because we are relying on the Netty's Channel API
-    var bytesRead = stream.read(buffer)
-    while (bytesRead >= 0) {
-      baos.write(buffer, 0, bytesRead)
-      bytesRead = stream.read(buffer)
+    val bytes = stream.read(buffer, 0, buffer.length) match {
+      case -1 =>
+        akka.util.ByteString.empty // EOF immediately
+      case n if n < buffer.length =>
+        akka.util.ByteString
+          .fromArrayUnsafe(buffer, 0, n) // Potentially wasteful, can we bet that the ByteString will be thrown away?
```

Comment:
@jrudolph WDYT about this?

Reply:
I don't think it should be a problem but I don't know the code paths in grpc exactly. @raboof what do you think?

Reply:
So, the trade-off is that for streams with unknown length, we can either:
* copy the bytes we need (allocation overhead)
* hold on to the complete 32k buffer (GC pressure)

The buffer is almost immediately passed to `protobufSerializer.deserialize`, so the question is whether the deserialized object is likely to keep a reference to the buffer. *currently* both for Java and Scala we use `CodedInputStream` configured to assume the buffer is mutable (so it will copy instead of wrap). That doesn't necessarily need to be the case in the future, though, of course.

So for now this looks like a safe improvement, though we might want to record why in a comment

Reply:
@raboof @jrudolph I'll add a comment. Alas, I am not sure we have any decent benchmarks exercising this part of the code.

Reply:
Comment added